File: src/main/java/me/rasztabiga/fridgy/products/domain/Product.java
package me.rasztabiga.fridgy.products.domain;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;
import javax.persistence.*;
import javax.validation.constraints.*;
import org.hibernate.annotations.Cache;
import org.hibernate.annotations.CacheConcurrencyStrategy;

/**
 * A Product.
 */
@Entity
@Table(name = "product")
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
public class Product implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "sequenceGenerator")
    @SequenceGenerator(name = "sequenceGenerator")
    private Long id;

    @NotNull
    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "ean_code")
    private String eanCode;

    @ManyToMany
    @Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
    @JoinTable(
        name = "rel_product__product_unit",
        joinColumns = @JoinColumn(name = "product_id"),
        inverseJoinColumns = @JoinColumn(name = "product_unit_id")
    )
    @JsonIgnoreProperties(value = { "products" }, allowSetters = true)
    private Set<ProductUnit> productUnits = new HashSet<>();

    @ManyToOne
    @JsonIgnoreProperties(value = { "products" }, allowSetters = true)
    private ProductCategory productCategory;

    // jhipster-needle-entity-add-field - JHipster will add fields here
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Product id(Long id) {
        this.id = id;
        return this;
    }

    public String getName() {
        return this.name;
    }

    public Product name(String name) {
        this.name = name;
        return this;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getEanCode() {
        return this.eanCode;
    }

    public Product eanCode(String eanCode) {
        this.eanCode = eanCode;
        return this;
    }

    public void setEanCode(String eanCode) {
        this.eanCode = eanCode;
    }

    public Set<ProductUnit> getProductUnits() {
        return this.productUnits;
    }

    public Product productUnits(Set<ProductUnit> productUnits) {
        this.setProductUnits(productUnits);
        return this;
    }

    public Product addProductUnit(ProductUnit productUnit) {
        this.productUnits.add(productUnit);
        productUnit.getProducts().add(this);
        return this;
    }

    public Product removeProductUnit(ProductUnit productUnit) {
        this.productUnits.remove(productUnit);
        productUnit.getProducts().remove(this);
        return this;
    }

    public void setProductUnits(Set<ProductUnit> productUnits) {
        this.productUnits = productUnits;
    }

    public ProductCategory getProductCategory() {
        return this.productCategory;
    }

    public Product productCategory(ProductCategory productCategory) {
        this.setProductCategory(productCategory);
        return this;
    }

    public void setProductCategory(ProductCategory productCategory) {
        this.productCategory = productCategory;
    }

    // jhipster-needle-entity-add-getters-setters - JHipster will add getters and setters here

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Product)) {
            return false;
        }
        return id != null && id.equals(((Product) o).id);
    }

    @Override
    public int hashCode() {
        // see https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/
        return getClass().hashCode();
    }

    // prettier-ignore
    @Override
    public String toString() {
        return "Product{" +
            "id=" + getId() +
            ", name='" + getName() + "'" +
            ", eanCode='" + getEanCode() + "'" +
            "}";
    }
}

File: src/main/kotlin/me/rasztabiga/fridgy/products/domain/Product.kt
package me.rasztabiga.fridgy.products.domain

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import org.hibernate.annotations.Cache
import org.hibernate.annotations.CacheConcurrencyStrategy
import java.io.Serializable
import javax.persistence.*
import javax.validation.constraints.*

/**
 * A Product.
 */
@Entity
@Table(name = "product")
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
data class Product(
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "sequenceGenerator")
    @SequenceGenerator(name = "sequenceGenerator")
    var id: Long? = null,
    @get: NotNull
    @Column(name = "name", nullable = false)
    var name: String? = null,

    @Column(name = "ean_code")
    var eanCode: String? = null,

    @ManyToMany
    @Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
    @JoinTable(
        name = "rel_product__product_unit",
        joinColumns = [
            JoinColumn(name = "product_id")
        ],
        inverseJoinColumns = [
            JoinColumn(name = "product_unit_id")
        ]
    )
    @JsonIgnoreProperties(
        value = [
            "products"
        ],
        allowSetters = true
    )
    var productUnits: MutableSet<ProductUnit>? = mutableSetOf(),

    @ManyToOne
    @JsonIgnoreProperties(
        value = [
            "products"
        ],
        allowSetters = true
    )
    var productCategory: ProductCategory? = null,

    // jhipster-needle-entity-add-field - JHipster will add fields here
) : Serializable {

    fun addProductUnit(productUnit: ProductUnit): Product {
        if (this.productUnits == null) {
            this.productUnits = mutableSetOf()
        }
        this.productUnits?.add(productUnit)
        productUnit.products?.add(this)
        return this
    }

    fun removeProductUnit(productUnit: ProductUnit): Product {
        this.productUnits?.remove(productUnit)
        return this
    }
    // jhipster-needle-entity-add-getters-setters - JHipster will add getters and setters here

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Product) return false

        return id != null && other.id != null && id == other.id
    }

    override fun hashCode() = 31

    override fun toString() = "Product{" +
        "id=$id" +
        ", name='$name'" +
        ", eanCode='$eanCode'" +
        "}"

    companion object {
        private const val serialVersionUID = 1L
    }
}
